### PR TITLE
Support for conditional upsert

### PR DESCRIPTION
**Pull request by prashant-shahi**
Adding support for conditional upsert.

---
This change is [Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph-js/63)

### CODE REVIEW

File: src/txn.ts [hunk 1]
```diff
@@ -125,20 +106,39 @@ export class Txn {
...
-            this.dc.debug(`Mutate request (ERR_FINISHED):\nmutation = ${stringifyMessage(mu)}`);
+            this.dc.debug(`Do request (ERR_FINISHED):\nquery = ${req.getQuery()}\nvars = ${req.getVarsMap()}`);
+            this.dc.debug(`Do request (ERR_FINISHED):\nmutation = ${stringifyMessage(req.getMutationsList()[0])}`);
             throw ERR_FINISHED;
         }
 
-        this.mutated = true;
-        mu.setStartTs(this.ctx.getStartTs());
-        this.dc.debug(`Mutate request:\n${stringifyMessage(mu)}`);
+        if (req.getMutationsList.length > 0) {
```
[the remark this answers is not in the document]
Comment (prashant-shahi) [reply src/txn.ts]: Oh right. Thanks!

**Review by paulftw — COMMENTED** [review]
> Reviewed 22 of 26 files at r1, 6 of 6 files at r2.
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph-js/63)* status: all files reviewed, 5 unresolved discussions (waiting on @mangalaman93, @prashant-shahi, and @pullrequest[bot])

___
*[protos/api.proto, line 170 at r2](https://reviewable.io/reviews/dgraph-io/dgraph-js/63#-LnmlnC83JFtOBsic3I6:-LnmlnC83JFtOBsic3I7:b-dnkv5d) ([raw file](https://github.com/dgraph-io/dgraph-js/blob/4c6535357a0fa6607b67120d102d181edd2a305b/protos/api.proto#L170)):*
> ```protobuf
> }
> 
> // vim: noexpandtab sw=2 ts=2
> ```

Just to confirm: this file is an identical copy of the api.proto in the main dgraph repo, right?

**Review by paulftw — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph-js/63)* status: all files reviewed, 6 unresolved discussions (waiting on @mangalaman93, @prashant-shahi, and @pullrequest[bot])

*[a discussion](https://reviewable.io/reviews/dgraph-io/dgraph-js/63#-LnmmjpH4653qgenauMo:-LnmmjpH4653qgenauMp:b-9fzi4r) (no related file):*
It seems like linReads is a separate unrelated API change, and if such it should be done in a separate PR.
OK to commit it as is since 1.1 release is so close, next time let's try to have a PR per feature.

Overall looks good to me.

---

**Review by prashant-shahi — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph-js/63)* status: all files reviewed, 6 unresolved discussions (waiting on @mangalaman93, @paulftw, and @pullrequest[bot])

*[a discussion](https://reviewable.io/reviews/dgraph-io/dgraph-js/63#-LnmmjpH4653qgenauMo:-Lnn0ezl0cPxnu9b7G8s:b-g2ysxo) (no related file):*
<details><summary><i>Previously, paulftw (Paul Korzhyk) wrote…</i></summary><blockquote>

It seems like linReads is a separate unrelated API change, and if such it should be done in a separate PR.
OK to commit it as is since 1.1 release is so close, next time let's try to have a PR per feature.

Overall looks good to me.
</blockquote></details>

oh, sure. will do it from now on.

---
___
*[protos/api.proto, line 18 at r1](https://reviewable.io/reviews/dgraph-io/dgraph-js/63#-LnmkdBWzoATVUA0_Ebl-r1-18:-Lnn0TTP1iSGuBtHB-xC:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph-js/blob/754ee707037b6e6e8222c70a732343be606d279f/protos/api.proto#L18)):*
<details><summary><i>Previously, pullrequest[bot] wrote…</i></summary><blockquote>

Camel case with an initial capital is known as Pascal case.
</blockquote></details>

Done.
___
*[protos/api.proto, line 19 at r1](https://reviewable.io/reviews/dgraph-io/dgraph-js/63#-LnmkdBWzoATVUA0_Ebl-r1-19:-Lnn0ToG-EOXoI9986g8:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph-js/blob/754ee707037b6e6e8222c70a732343be606d279f/protos/api.proto#L19)):*
<details><summary><i>Previously, pullrequest[bot] wrote…</i></summary><blockquote>

underscore_separated_names is known as Snake case
</blockquote></details>

Done.
___
*[protos/api.proto, line 26 at r1](https://reviewable.io/reviews/dgraph-io/dgraph-js/63#-LnmkdBWzoATVUA0_Ebl-r1-26:-Lnn0UGV9cBeupAb8max:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph-js/blob/754ee707037b6e6e8222c70a732343be606d279f/protos/api.proto#L26)):*
<details><summary><i>Previously, pullrequest[bot] wrote…</i></summary><blockquote>

Do these commented out lines still need to be part of the code? If yes - probably make them all one multiline comment, or multiple single-line comments. 

Right now there are multiple multiline comments of 1 line each.
</blockquote></details>

Done.
___
*[protos/api.proto, line 170 at r2](https://reviewable.io/reviews/dgraph-io/dgraph-js/63#-LnmlnC83JFtOBsic3I6:-Lnn0UxJ-rCa59qI9l2B:b-pfmebw) ([raw file](https://github.com/dgraph-io/dgraph-js/blob/4c6535357a0fa6607b67120d102d181edd2a305b/protos/api.proto#L170)):*
<details><summary><i>Previously, paulftw (Paul Korzhyk) wrote…</i></summary><blockquote>

Just to confirm: this file is an identical copy of the api.proto in the main dgraph repo, right?
</blockquote></details>

Yes. It is identical. Just that few comments have been either altered and removed.

**Pull request merged**